app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt

```diff
@@ -995,6 +998,102 @@ class StateFragmentLocalTest {
     }
   }
 
+  @Test
+  @Config(qualifiers = "port")
+  fun testStateFragment_mobilePortrait_finishExploration_endOfSessionConfettiIsDisplayed() {
+    launchForExploration(FRACTIONS_EXPLORATION_ID_1).use {
+      startPlayingExploration()
+      playThroughAllStates()
+      clickContinueButton()
+
+      onView(withId(R.id.full_screen_confetti_view)).check(matches(hasActiveConfetti()))
+    }
+  }
+
+  @Test
+  @Config(qualifiers = "land")
+  fun testStateFragment_mobileLandscape_finishExploration_endOfSessionConfettiIsDisplayed() {
+    launchForExploration(FRACTIONS_EXPLORATION_ID_1).use {
+      startPlayingExploration()
+      playThroughAllStates()
+      clickContinueButton()
+
+      onView(withId(R.id.full_screen_confetti_view)).check(matches(hasActiveConfetti()))
+    }
+  }
+
+  @Test
+  @Ignore("Currently failing as tablet layouts are not properly rendering interaction views")
+  @Config(qualifiers = "sw600dp-port")
+  fun testStateFragment_tabletPortrait_finishExploration_endOfSessionConfettiIsDisplayed() {
+    launchForExploration(FRACTIONS_EXPLORATION_ID_1).use {
+      startPlayingExploration()
+      playThroughAllStates()
+      clickContinueButton()
+
+      onView(withId(R.id.full_screen_confetti_view)).check(matches(hasActiveConfetti()))
+    }
+  }
+
+  @Test
+  @Ignore("Currently failing as tablet layouts are not properly rendering interaction views")
+  @Config(qualifiers = "sw600dp-land")
+  fun testStateFragment_tabletLandscape_finishExploration_endOfSessionConfettiIsDisplayed() {
+    launchForExploration(FRACTIONS_EXPLORATION_ID_1).use {
+      startPlayingExploration()
+      playThroughAllStates()
+      clickContinueButton()
+
+      onView(withId(R.id.full_screen_confetti_view)).check(matches(hasActiveConfetti()))
+    }
+  }
+
+  @Test
+  fun testStateFragment_submitCorrectAnswer_endOfSessionConfettiDoesNotStart() {
+    launchForExploration(FRACTIONS_EXPLORATION_ID_1).use {
+      startPlayingExploration()
+      playThroughState1()
+
+      onView(withId(R.id.full_screen_confetti_view)).check(matches(not(hasActiveConfetti())))
+    }
+  }
+
+  @Test
+  fun testStateFragment_notAtEndOfExploration_endOfSessionConfettiDoesNotStart() {
+    launchForExploration(FRACTIONS_EXPLORATION_ID_1).use {
+      startPlayingExploration()
+      // Play through all questions but do not reach the last screen of the exploration.
+      playThroughAllStates()
+
+      onView(withId(R.id.full_screen_confetti_view)).check(matches(not(hasActiveConfetti())))
+    }
+  }
+
+  @Test
+  fun testStateFragment_reachEndOfExplorationTwice_endOfSessionConfettiIsDisplayedOnce() {
+    launchForExploration(FRACTIONS_EXPLORATION_ID_1).use {
+      startPlayingExploration()
+      playThroughAllStates()
+      clickContinueButton()
+      onView(withId(R.id.full_screen_confetti_view)).check(matches(hasActiveConfetti()))
+      onView(withId(R.id.full_screen_confetti_view)).check(
```

Comment:
Is the intention of this test to navigate back during an active confetti view? If so, I suggest also adding a test that navigates away & back to the end state after the previous confetti finished to verify it replays from scratch (as designed).

Reply:
> Is the intention of this test to navigate back during an active confetti view?

Yes!

> If so, I suggest also adding a test that navigates away & back to the end state after the previous confetti finished to verify it replays from scratch (as designed).

Good point -- having trouble running the animation to completion after starting it since Robolectric doesn't call onDraw() for rendering. The animations are time-based (they finish after 5000ms) but running the coroutine dispatchers doesn't appear to run through the animations, presumably because Robolectric doesn't render them. But not sure how force the animation to completion as calling `invalidate()` and `draw(Canvas())` doesn't call the KonfettiView's onDraw() function (can't call `onDraw()` directly on the view since it's protected)

Reply:
Got it. I suspect it isn't possible to test this, then (or its pretty hard & involves custom shadows).